File: src/hub_catalog.rs
//! Hub repository catalog.
//! Defines the repositories that are currently managed by the hub and
//! repositories planned as future dependency targets.

use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RepositoryStage {
    Active,
    Future,
}

#[derive(Debug, Clone, Copy)]
pub struct RepositoryDefinition {
    pub canonical_name: &'static str,
    pub stage: RepositoryStage,
    pub aliases: &'static [&'static str],
}

pub const HUB_REPOSITORIES: [RepositoryDefinition; 5] = [
    RepositoryDefinition {
        canonical_name: "RyzeSpace.Client",
        stage: RepositoryStage::Active,
        aliases: &[
            "client",
            "ryzespace-client",
            "ryzespace_client",
            "ryzespace/client",
            "@ryzespace/client",
        ],
    },
    RepositoryDefinition {
        canonical_name: "RyzeSpace.HelpCenter",
        stage: RepositoryStage::Active,
        aliases: &[
            "helpcenter",
            "help-center",
            "ryzespace-helpcenter",
            "ryzespace_helpcenter",
            "ryzespace/helpcenter",
            "@ryzespace/helpcenter",
        ],
    },
    RepositoryDefinition {
        canonical_name: "RyzeSpace.AdminPanel",
        stage: RepositoryStage::Future,
        aliases: &[
            "admin",
            "admin-panel",
            "panel-admina",
            "ryzespace-adminpanel",
            "ryzespace/admin-panel",
            "@ryzespace/admin-panel",
        ],
    },
    RepositoryDefinition {
        canonical_name: "RyzeSpace.Mobile",
        stage: RepositoryStage::Future,
        aliases: &[
            "mobile",
            "app-mobile",
            "aplikacja-mobilna",
            "ryzespace-mobile",
            "ryzespace/mobile",
            "@ryzespace/mobile",
        ],
    },
    RepositoryDefinition {
        canonical_name: "RyzeSpace.Desktop",
        stage: RepositoryStage::Future,
        aliases: &[
            "desktop",
            "desktop-app",
            "aplikacja-desktopowa",
            "ryzespace-desktop",
            "ryzespace/desktop",
            "@ryzespace/desktop",
        ],
    },
];

pub fn active_repository_names() -> Vec<String> {
    repository_names_by_stage(RepositoryStage::Active)
}

pub fn future_repository_names() -> Vec<String> {
    repository_names_by_stage(RepositoryStage::Future)
}

pub fn tracked_repository_names() -> Vec<String> {
    HUB_REPOSITORIES
        .iter()
        .map(|repo| repo.canonical_name.to_string())
        .collect()
}

pub fn repository_names_by_stage(stage: RepositoryStage) -> Vec<String> {
    HUB_REPOSITORIES
        .iter()
        .filter(|repo| repo.stage == stage)
        .map(|repo| repo.canonical_name.to_string())
        .collect()
}

pub fn resolve_canonical_repository_name(value: &str) -> Option<&'static str> {
    let normalized_value = normalize_repository_token(value);
    if normalized_value.is_empty() {
        return None;
    }

    HUB_REPOSITORIES
        .iter()
        .find(|repo| repository_tokens(repo).contains(&normalized_value))
        .map(|repo| repo.canonical_name)
}

pub fn dependency_identifiers(value: &str) -> Vec<String> {
    if let Some(repo) = find_repository_definition(value) {
        let mut identifiers = Vec::new();
        identifiers.push(repo.canonical_name.to_string());
        identifiers.extend(repo.aliases.iter().map(|alias| alias.to_string()));
        identifiers.extend(derived_identifiers(repo.canonical_name));
        dedupe_preserve_order(identifiers)
    } else {
        dedupe_preserve_order(derived_identifiers(value))
    }
}

pub fn normalize_repository_token(value: &str) -> String {
    value
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .map(|c| c.to_ascii_lowercase())
        .collect()
}

fn find_repository_definition(value: &str) -> Option<&'static RepositoryDefinition> {
    let normalized_value = normalize_repository_token(value);
    HUB_REPOSITORIES
        .iter()
        .find(|repo| repository_tokens(repo).contains(&normalized_value))
}

fn repository_tokens(repo: &RepositoryDefinition) -> HashSet<String> {
    let mut tokens = HashSet::new();
    tokens.insert(normalize_repository_token(repo.canonical_name));
    for alias in repo.aliases {
        tokens.insert(normalize_repository_token(alias));
    }
    for identifier in derived_identifiers(repo.canonical_name) {
        tokens.insert(normalize_repository_token(&identifier));
    }
    tokens
}

fn derived_identifiers(value: &str) -> Vec<String> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return Vec::new();
    }

    let mut identifiers = vec![trimmed.to_string(), trimmed.to_lowercase()];
    let parts: Vec<&str> = trimmed
        .split(|c: char| matches!(c, '.' | '-' | '_' | '/' | '@'))
        .filter(|part| !part.is_empty())
        .collect();

    if let Some(last) = parts.last() {
        identifiers.push((*last).to_string());
        identifiers.push(last.to_lowercase());
    }

    if parts.len() >= 2 {
        let namespace = parts[..parts.len() - 1].join("");
        let namespace_kebab = parts[..parts.len() - 1]
            .iter()
            .map(|part| part.to_ascii_lowercase())
            .collect::<Vec<_>>()
            .join("-");
        let name = parts[parts.len() - 1].to_ascii_lowercase();

        identifiers.push(format!("{}.{}", namespace, name));
        identifiers.push(format!("{}-{}", namespace, name));
        identifiers.push(format!("{}_{}", namespace, name));
        identifiers.push(format!("{}/{}", namespace, name));
        identifiers.push(format!("@{}/{}", namespace_kebab, name));
    }

    dedupe_preserve_order(identifiers)
}

fn dedupe_preserve_order(values: Vec<String>) -> Vec<String> {
    let mut seen = HashSet::new();
    let mut result = Vec::new();

    for value in values {
        if value.trim().is_empty() {
            continue;
        }

        let key = normalize_repository_token(&value);
        if seen.insert(key) {
            result.push(value);
        }
    }

    result
}

File: src/RyzeHub.Application/Hub/HubCatalog.cs
namespace RyzeHub.Application.Hub;

public enum RepositoryStage
{
    Active,
    Future
}

public sealed record RepositoryDefinition(string CanonicalName, RepositoryStage Stage, IReadOnlyList<string> Aliases);

/// <summary>
/// Repositories managed by the hub plus the planned dependency targets.
/// </summary>
public static class HubCatalog
{
    public static IReadOnlyList<RepositoryDefinition> Repositories { get; } =
    [
        new("RyzeAuth", RepositoryStage.Active,
            ["ryzeauth", "ryze-auth", "ryzespace/ryzeauth", "@ryzespace/auth", "RyzeSpace.Auth"]),
        new("RyzeSpace.Client", RepositoryStage.Active,
            ["client", "ryzespace-client", "ryzespace_client", "ryzespace/client", "@ryzespace/client"]),
        new("RyzeSpace.HelpCenter", RepositoryStage.Active,
            ["helpcenter", "help-center", "ryzespace-helpcenter", "ryzespace_helpcenter", "ryzespace/helpcenter", "@ryzespace/helpcenter"]),
        new("RyzeSpace.AdminPanel", RepositoryStage.Future,
            ["admin", "admin-panel", "panel-admina", "ryzespace-adminpanel", "ryzespace/admin-panel", "@ryzespace/admin-panel"]),
        new("RyzeSpace.Mobile", RepositoryStage.Future,
            ["mobile", "app-mobile", "aplikacja-mobilna", "ryzespace-mobile", "ryzespace/mobile", "@ryzespace/mobile"]),
        new("RyzeSpace.Desktop", RepositoryStage.Future,
            ["desktop", "desktop-app", "aplikacja-desktopowa", "ryzespace-desktop", "ryzespace/desktop", "@ryzespace/desktop"])
    ];

    public static IReadOnlyList<string> ActiveRepositoryNames() => NamesByStage(RepositoryStage.Active);

    public static IReadOnlyList<string> FutureRepositoryNames() => NamesByStage(RepositoryStage.Future);

    public static IReadOnlyList<string> TrackedRepositoryNames() =>
        [.. Repositories.Select(repository => repository.CanonicalName)];

    public static IReadOnlyList<string> NamesByStage(RepositoryStage stage) =>
        [.. Repositories.Where(repository => repository.Stage == stage).Select(repository => repository.CanonicalName)];

    public static string? ResolveCanonicalRepositoryName(string value)
    {
        var normalized = NormalizeRepositoryToken(value);
        if (normalized.Length == 0)
        {
            return null;
        }

        return Repositories
            .FirstOrDefault(repository => RepositoryTokens(repository).Contains(normalized))
            ?.CanonicalName;
    }

    public static IReadOnlyList<string> DependencyIdentifiers(string value)
    {
        var definition = FindRepositoryDefinition(value);
        if (definition is null)
        {
            return DedupePreserveOrder(DerivedIdentifiers(value));
        }

        var identifiers = new List<string> { definition.CanonicalName };
        identifiers.AddRange(definition.Aliases);
        identifiers.AddRange(DerivedIdentifiers(definition.CanonicalName));
        return DedupePreserveOrder(identifiers);
    }

    public static string NormalizeRepositoryToken(string value) =>
        new([.. value.Where(char.IsAsciiLetterOrDigit).Select(char.ToLowerInvariant)]);

    private static RepositoryDefinition? FindRepositoryDefinition(string value)
    {
        var normalized = NormalizeRepositoryToken(value);
        return Repositories.FirstOrDefault(repository => RepositoryTokens(repository).Contains(normalized));
    }

    private static HashSet<string> RepositoryTokens(RepositoryDefinition repository)
    {
        var tokens = new HashSet<string>(StringComparer.Ordinal)
        {
            NormalizeRepositoryToken(repository.CanonicalName)
        };

        foreach (var alias in repository.Aliases)
        {
            tokens.Add(NormalizeRepositoryToken(alias));
        }

        foreach (var identifier in DerivedIdentifiers(repository.CanonicalName))
        {
            tokens.Add(NormalizeRepositoryToken(identifier));
        }

        return tokens;
    }

    private static List<string> DerivedIdentifiers(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length == 0)
        {
            return [];
        }

        var identifiers = new List<string> { trimmed, trimmed.ToLowerInvariant() };
        var parts = trimmed.Split(['.', '-', '_', '/', '@'], StringSplitOptions.RemoveEmptyEntries);

        if (parts.Length > 0)
        {
            identifiers.Add(parts[^1]);
            identifiers.Add(parts[^1].ToLowerInvariant());
        }

        if (parts.Length >= 2)
        {
            var namespacePart = string.Concat(parts[..^1]);
            var namespaceKebab = string.Join('-', parts[..^1].Select(part => part.ToLowerInvariant()));
            var name = parts[^1].ToLowerInvariant();

            identifiers.Add($"{namespacePart}.{name}");
            identifiers.Add($"{namespacePart}-{name}");
            identifiers.Add($"{namespacePart}_{name}");
            identifiers.Add($"{namespacePart}/{name}");
            identifiers.Add($"@{namespaceKebab}/{name}");
        }

        return DedupePreserveOrder(identifiers);
    }

    private static List<string> DedupePreserveOrder(IEnumerable<string> values)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var result = new List<string>();

        foreach (var value in values)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                continue;
            }

            if (seen.Add(NormalizeRepositoryToken(value)))
            {
                result.Add(value);
            }
        }

        return result;
    }
}
